### ISSUE TITLE
More help on vouch-proxy  - doGetAuthenticationInfo failed: Unexpected response from server: 401 / Unauthorized after 5 minutes

### ISSUE THREAD

**Issue by IMbiea** [flytreeleft/nexus3-keycloak-plugin#53]
Hello!  Thanks for the help before on vouch proxy.   We're finally getting around to getting it going.

A few things to mention beforehand.

1) using nginx-ingress-controller
2) using voucher 17.3-alpine
3) kubernetes 1.18

5 minute snapshot is:

```
2020-10-07 19:53:47,689+0000 INFO  [qtp330456734-217] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:53:47,690+0000 INFO  [qtp330456734-217] test-user org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthorizationInfo for test-user
2020-10-07 19:53:47,776+0000 INFO  [qtp330456734-216] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:53:47,963+0000 INFO  [qtp330456734-213] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:53:48,079+0000 INFO  [qtp330456734-219] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:53:48,098+0000 INFO  [qtp330456734-218] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:53:48,110+0000 INFO  [qtp330456734-225] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:53:48,117+0000 INFO  [qtp330456734-46] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:53:48,124+0000 INFO  [qtp330456734-217] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:54:48,097+0000 INFO  [qtp330456734-46] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:24,706+0000 INFO  [qtp330456734-46] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:24,804+0000 INFO  [qtp330456734-216] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:25,083+0000 INFO  [qtp330456734-218] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:25,112+0000 INFO  [qtp330456734-217] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:25,123+0000 INFO  [qtp330456734-219] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:25,124+0000 INFO  [qtp330456734-213] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:25,511+0000 INFO  [qtp330456734-225] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:25,511+0000 INFO  [qtp330456734-46] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:49,058+0000 INFO  [qtp330456734-216] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:55:49,060+0000 INFO  [qtp330456734-216] test-user org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthorizationInfo for test-user
2020-10-07 19:56:25,100+0000 INFO  [qtp330456734-216] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:56:49,053+0000 INFO  [qtp330456734-216] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:25,102+0000 INFO  [qtp330456734-218] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:50,057+0000 INFO  [qtp330456734-218] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:50,059+0000 INFO  [qtp330456734-218] test-user org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthorizationInfo for test-user
2020-10-07 19:57:56,761+0000 INFO  [qtp330456734-225] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:56,768+0000 INFO  [qtp330456734-219] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:56,774+0000 INFO  [qtp330456734-213] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:56,783+0000 INFO  [qtp330456734-217] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:56,788+0000 INFO  [qtp330456734-46] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:56,792+0000 INFO  [qtp330456734-216] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:56,991+0000 INFO  [qtp330456734-232] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:57:57,017+0000 INFO  [qtp330456734-218] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:58:25,103+0000 INFO  [qtp330456734-217] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:58:38,926+0000 INFO  [qtp330456734-218] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:58:38,932+0000 INFO  [qtp330456734-46] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo for test-user via org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakHttpHeaderAuthToken: true
2020-10-07 19:58:51,054+0000 INFO  [qtp330456734-216] *UNKNOWN org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm - doGetAuthenticationInfo failed: Unexpected response from server: 401 / Unauthorized
org.github.flytreeleft.nexus3.keycloak.plugin.internal.http.HttpResponseException: Unexpected response from server: 401 / Unauthorized
        at org.github.flytreeleft.nexus3.keycloak.plugin.internal.http.HttpMethod.execute(HttpMethod.java:65)
        at org.github.flytreeleft.nexus3.keycloak.plugin.internal.http.HttpMethodResponse$2.execute(HttpMethodResponse.java:37)
        at org.github.flytreeleft.nexus3.keycloak.plugin.internal.KeycloakAdminClient.obtainUserInfo(KeycloakAdminClient.java:88)
        at org.github.flytreeleft.nexus3.keycloak.plugin.internal.NexusKeycloakClient.authenticate(NexusKeycloakClient.java:70)
        at org.github.flytreeleft.nexus3.keycloak.plugin.internal.NexusKeycloakClient.authenticate(NexusKeycloakClient.java:53)
        at org.github.flytreeleft.nexus3.keycloak.plugin.KeycloakAuthenticatingRealm.doGetAuthenticationInfo(KeycloakAuthenticatingRealm.java:93)
        at org.apache.shiro.realm.AuthenticatingRealm.getAuthenticationInfo(AuthenticatingRealm.java:571)
        at org.sonatype.nexus.security.authc.FirstSuccessfulModularRealmAuthenticator.doMultiRealmAuthentication(FirstSuccessfulModularRealmAuthenticator.java:59)
        at org.apache.shiro.authc.pam.ModularRealmAuthenticator.doAuthenticate(ModularRealmAuthenticator.java:275)
        at org.apache.shiro.authc.AbstractAuthenticator.authenticate(AbstractAuthenticator.java:198)
        at org.apache.shiro.mgt.AuthenticatingSecurityManager.authenticate(AuthenticatingSecurityManager.java:106)
```

vouch-config:

```yaml
# vouch config
config:
  vouch:
    allowAllUsers: false
    domains:
    - test.com
    testing: false
    logLevel: debug
    jwt:
      issuer: TeSt
      maxAge: 14400
    cookie:
      maxAge: 0
    headers:
      jwt: X-Vouch-Token
      querystring: access_token
      redirect: X-Vouch-Requested-URI
      claims:
        - preferred_username
      accesstoken: X-Vouch-IdP-AccessToken
  oauth:
    provider: oidc
    client_id: nexus
    client_secret: 794bd70f-4012-4614-b9ff-dc1cb8943341
    callback_url: https://voucher-tst-test-test.com/auth
    auth_url: https://keycloak-tst-test-test.com/auth/realms/TeSt-tst/protocol/openid-connect/auth
    token_url: https://keycloak-tst-test-test.com/auth/realms/TeSt-tst/protocol/openid-connect/token
    user_info_url: https://keycloak-tst-test-test.com/auth/realms/TeSt-tst/protocol/openid-connect/userinfo
    end_session_endpoint: https://keycloak-tst-test-test.com/auth/realms/TeSt-tst/protocol/openid-connect/logout
    scopes:
      - openid
      - email
      - profile
    preferredDomain:

# existingSecretName -- Allow overriding the config value with an existing secret, like a sealed secret
```

I've been playing with keycloak configs.. but it's just not consistant

**Comment by flytreeleft** [comment]
@IMbiea Did you check whether the access token is invalid or not after 5 min? 

Please enable the debug log for Keycloak with setting the environment `KEYCLOAK_LOGLEVEL` to `DEBUG`, and paste the logs after the 401 arising again.

**Comment by IMbiea** [comment]
@flytreeleft , yes when I get the 401 on nexus, the keycloak logs show:

`17:46:19,260 WARN  [org.keycloak.events] (default task-80) type=USER_INFO_REQUEST_ERROR, realmId=TeSt-tst, clientId=null, userId=null, ipAddress=192.168.5.72, error=invalid_token, auth_method=validate_access_token`

And then I just get tons of pops ups asking to login via the nexus gui, but never a re-direct back to keycloak/vouch

**Comment by IMbiea** [comment]
I wonder... do I need additional mappers in keycloak?
![image](https://user-images.githubusercontent.com/27780437/95795427-8b3c7900-0ca7-11eb-92d8-4445d9f84c36.png)

**Comment by IMbiea** [comment]
I'm a bit ignorant on this.. but did find this: https://github.com/jhipster/generator-jhipster/issues/6929, but uncertain how to apply this logic in either nexus or voucher.

**Comment by IMbiea** [comment]
:-(  I'm now wondering if it's the lack of /oauth/expired in vouch?

is there a work around for this?

https://github.com/vouch/vouch-proxy/issues/149
and
https://github.com/vouch/vouch-proxy/issues/128

**Comment by flytreeleft** [comment]
@IMbiea Sorry, I'm busy working on my job. I'll check this when I have free time. :)

**Comment by flytreeleft** [comment]
Please try the answer in the issue [#54](https://github.com/flytreeleft/nexus3-keycloak-plugin/issues/54#issuecomment-710840426).

**Comment by IMbiea** [comment]
as soon as I can figure out the right nginx-ingress syntax  :-D   :-(

**Comment by flytreeleft** [comment]
The issue may be fixed in #54 already, close it.
